sympy/map/map.py

```diff
@@ -0,0 +1,193 @@
+from sympy.core import Expr, S, Tuple
+from sympy.core.compatibility import is_sequence, ordered
+from sympy.core.function import arity
+from sympy.core.sympify import _sympify
+from sympy.core.symbol import Str
+from sympy.sets.sets import FiniteSet
+
+class Map(Expr):
+    """
+    A class for general mathematical mappings
+
+    Parameters
+    ==========
+
+    parameters : tuple
+        Parameters of the map
+
+    domain, codomain : Set
+
+    name : str
+        Name of the map
+
+    Examples
+    ========
+
+    >>> from sympy import symbols
+    >>> from sympy.map import Map
+    >>> x, y, z = symbols('x y z')
+
+    >>> f = Map(parameters=(x, ), name='f')
+    >>> f.name
+    'f'
+    >>> f.parameters
+    (x,)
+    >>> f.nargs
+    Naturals0
+
+    >>> class G(Map):
+    ...     def eval(self, a, b):
+    ...         x,y = self.parameters
+    ...         return a*x + b*y
+    >>> g = G(parameters=(x,y), name='g')
+
+    >>> g.name
+    'g'
+    >>> g.parameters
+    (x, y)
+    >>> g.nargs
+    FiniteSet(2)
+
+    >>> g.subs(x,2).parameters
+    (2, y)
+    >>> g.subs(y, 4).parameters
+    (x, 4)
+
+    """
+    def __new__(
+        cls, parameters=(), domain=S.UniversalSet, codomain=S.UniversalSet,
+        name='f', **kwargs
+    ):
+        parameters = Tuple(*[_sympify(a) for a in parameters])
+
+        domain, codomain = _sympify(domain), _sympify(codomain)
+
+        if not isinstance(name, Str):
+            name = Str(name)
+
+        nargs = cls._find_nargs()
+
+        obj = super().__new__(cls, parameters, domain, codomain, name)
+        obj._nargs = nargs
+        return obj
+
+    @property
+    def parameters(self):
+        return self.args[0]
+
+    @property
+    def domain(self):
+        return self.args[1]
+
+    @property
+    def codomain(self):
+        return self.args[2]
+
+    @property
+    def name(self):
+        return self.args[3]
+
+    @classmethod
+    def _find_nargs(cls):
```

Comment:
I thought that the number of arguments can be matched to the number of cartesian products of the domain

Reply:
Probably worth looking at how `Lambda/ImageSet` handles this including for nested tuples. A multi-arg Lambda only works when multiple base sets are given to ImageSet. Otherwise ProductSets can be unpacked using a nested tuple signature:
```julia
In [9]: A = FiniteSet(1, 2)                                                                                            

In [10]: B = FiniteSet(3, 4)                                                                                           

In [11]: ImageSet(Lambda((x, y), x+y), A, B)                                                                           
Out[11]: {x + y | x ∊ {1, 2} , y ∊ {3, 4}}

In [12]: ImageSet(Lambda(((x, y),), x+y), A * B)                                                                       
Out[12]: {x + y | (x, y) ∊ {1, 2} × {3, 4}}
```

Reply:
I applied it in commit `d43072a`. How is it now?